ML-DSA private key details

jakemas opened the pull request:
## Changes to Support Multiple Private Key Formats

This PR updates the specification and examples to support three different private key formats as defined in the ASN.1 module:

## Key Changes
- updated public key section 4 to no longer describe private keys `ML-DSA-XX-PublicKey`
- updated private key section 6 to provide description of `ML-DSA-XX-PrivateKey` e.g.
```
ML-DSA-65-PrivateKey ::= CHOICE {
  seed [0] OCTET STRING (SIZE (32)),
  expandedKey OCTET STRING (SIZE (4032)),
  both SEQUENCE {
      seed OCTET STRING (SIZE (32)),
      expandedKey OCTET STRING (SIZE (4032))
      }
  }
```
- updated `generate.go` source code for new examples (seed, expanded, both)
- added new examples (and ASN.1 pretty print)

## Examples

Each security level (ML-DSA-44, ML-DSA-65, ML-DSA-87) now includes examples of all three formats, showing:
- How the same seed produces different expanded keys
- ASN.1 structure and encoding

## Implementation Notes

The PR includes updates to the example generation code to support all formats and ensure consistent encoding across implementations.

jakemas commented:
why is only 1 workflow running?

jakemas commented:
Should we tag the expandedKey rather than the seed?

```
ML-DSA-65-PrivateKey ::= CHOICE {
  seed OCTET STRING (SIZE (32)),
  expandedKey [0] OCTET STRING (SIZE (4032)),
  both SEQUENCE {
      seed OCTET STRING (SIZE (32)),
      expandedKey OCTET STRING (SIZE (4032))
      }
  }
```

It saves encoding on the seed.

vdukhovni commented:
> Should we tag the expandedKey rather than the seed?
> 
> ```
> ML-DSA-65-PrivateKey ::= CHOICE {
>   seed OCTET STRING (SIZE (32)),
>   expandedKey [0] OCTET STRING (SIZE (4032)),
>   both SEQUENCE {
>       seed OCTET STRING (SIZE (32)),
>       expandedKey OCTET STRING (SIZE (4032))
>       }
>   }
> ```
> 
> It saves total seed encoding size, i.e. the compressed format.

The tags are **IMPLICIT**, which side has the tag has no impact on size. **ALL** it does is change the tag from `0x04` to `0x80`, the rest of the data is completely unchanged.
Since having the expanded key be an OCTET STRING (rather than IMPLICIT OCTET STRING) interoperates with OQS, it is best to NOT switch the tags.

In summary, let's NOT make the switch.

dghgit commented:
There was a discussion early on about tagging the expanded key instead, but tagging the seed means it's compatible with what we were all shipping prior to IETF Dublin 2024 - Bouncy Castle was also using the expanded key as it was used in the earlier drafts. As Viktor points out, the tag's implicit, so no cost - the expanded-key and seed strings are both "optimal".

So, yes, tag the seed, no-one's had a great time with the private key encoding, but if we actually manage to emerge with something that's backwards compatible, at least we can finish this knowing some people are happy.

jakemas commented:
> > Should we tag the expandedKey rather than the seed?
> > ```
> > ML-DSA-65-PrivateKey ::= CHOICE {
> >   seed OCTET STRING (SIZE (32)),
> >   expandedKey [0] OCTET STRING (SIZE (4032)),
> >   both SEQUENCE {
> >       seed OCTET STRING (SIZE (32)),
> >       expandedKey OCTET STRING (SIZE (4032))
> >       }
> >   }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It saves total seed encoding size, i.e. the compressed format.
> 
> The tags are **IMPLICIT**, which side has the tag has no impact on size. **ALL** it does is change the tag from `0x04` to `0x80`, the rest of the data is completely unchanged. Since having the expanded key be an OCTET STRING (rather than IMPLICIT OCTET STRING) interoperates with OQS, it is best to NOT switch the tags.
> 
> In summary, let's NOT make the switch.

Thank you! Came to this realisation myself today also - agreed. Keep it as it is.

jakemas commented:
>There was a discussion early on about tagging the expanded key instead, but tagging the seed means it's compatible with what we were all shipping prior to IETF Dublin 2024 - Bouncy Castle was also using the expanded key as it was used in the earlier drafts. As Viktor points out, the tag's implicit, so no cost - the expanded-key and seed strings are both "optimal".

Completely agree here, thank you for the confirmation -- you can see what I was thinking with trying to preserve backwards compatibility with previous versions of the draft here by moving the tag from the seed.

jakemas commented:
Thank you both [dghgit](https://github.com/dghgit) and [vdukhovni](https://github.com/vdukhovni) for taking a look here! I wanted to get both your feedback on the proposed private key format.

vdukhovni commented:
> Thank you both [dghgit](https://github.com/dghgit) and [vdukhovni](https://github.com/vdukhovni) for taking a look here! I wanted to get both your feedback on the proposed private key format.

Many thanks for the prompt fix. I have confirmed that the new files are identical to those I generate from the same seed for all three choices and all three parameter sets.

vdukhovni commented on draft-ietf-lamps-dilithium-certificates.md:
```diff
@@ -354,8 +357,9 @@ encoding asymmetric keypairs. When an ML-DSA private key or keypair is encoded a
-the ML-DSA-PrivateKey type.
+environment that uses ASN.1 encoding, it can be encoded using one of the
+the ML-DSA-PrivateKey CHOICE formats defined in {{priv-key}}. The seed format
+is RECOMMENDED as it efficiently stores both the private and public key.
```
While the ML-DSA private key does not contain a copy of the public key, the latter can be computed from the private key even in the absence of a seed. So the last sentence is misleading. It seems to suggest that the expanded key would require providing the public key separately, which is not the case.

Computing the public key from the seed entails:

1. Generating the private key s1 and s2 vectors and matrix from the seed.
2. Computing the public key from the s1, s2 and the matrix.

While with the expanded key, one need only compute the matrix in step 1.

So the "public" key is irrelevant here, or else the expanded key actually has an advantage!
The seed is of course more space efficient vs. the expanded key for representing the private key, and one may be more confident that `s1` and `s2` are validly generated, ... but I don't think the reason given for the preference holds up.

vdukhovni replied:
Well, the text seems to say that the seed is recommended **because**:
```
it efficiently stores both the private and public key
```
But this is NOT a reason to prefer the seed, because the expanded key also stores "both the private and public key". If the reason is just that it is smaller than than the expanded key, that should be reason. If the reason is about being sure that the private vectors are valid outputs of keygen, then that should be the reason, ... but avoiding the need to separately provide a public key is NOT a reason that the seed is preferred over `sk`.

vdukhovni commented on draft-ietf-lamps-dilithium-certificates.md:
```diff
@@ -431,10 +432,72 @@ OneAsymmetricKey is replicated below.
+the appropriate OID as defined in Section 3. If present, the "publicKey"
+field will hold the encoded public key as defined in Section 5.
+
+NOTE: While the private key can be stored in multiple formats, the seed-only
+format is RECOMMENDED as it is the most compact representation. Both the
+expanded private key and the public key can be deterministically derived
+from the seed using ML-DSA.KeyGen_internal(xi). While the publicKey field
+and expandedKey format are technically redundant when using the seed-only
+format, they MAY be included to enable keypair consistency checks during
+import operations.
```
Same issue here.  The first sentence is correct (more compact representation), but what follows again suggests that the public key can only be recovered from the seed, which is NOT the case. OpenSSL, for example, always recovers the public key from the private alone when importing ML-DSA private keys (if a public key is also supplied on import it is compared with the generated value and the pair reject if it does not match).

jakemas replied:
Agree here also, similarly, AWS-LC does the same, we generate the public key based on the provided private key. Again, the sentence you interpret differently to me is simply pointed out that the seed _can_ provide both pub and priv, not that it can only be recovered.

I was actually going to propose a section in the draft that speaks to how to generate pk from sk, given that this isn't provided in the reference implementation. I mentioned that in this issue: https://github.com/lamps-wg/dilithium-certificates/issues/77. Perhaps such an inclusion can then be referred to in the text you point out here and above, to indicate that the pk can be constructed from the sk. I think that should help satisfy your remarks.

vdukhovni replied:
Well, just a subset (4–8) of the Algorithm 6 ML-DSA.KeyGen_internal(𝜉) steps, since the private vectors are generated, and then the public key just from those.

vdukhovni reviewed (commented):
My comments are largely editorial nits based on availability of private -> public key recovery in ML-DSA.

seanturner commented:
This one seems to loose the pretty print. Should we add those back in?

jakemas commented:
> This one seems to loose the pretty print. Should we add those back in?

Thanks Sean! Great catch, updated `pretty` to generate these new pretty print examples, and updated make file.

jakemas commented:
Thanks Sean! Added pretty print back in! What do you think of the section?

We can still iterate on exact language, but I’d love to merge to update the working draft to show the intension of the changes.

Cheers,
Jake

From: Sean Turner ***@***.***>
Reply-To: lamps-wg/dilithium-certificates ***@***.***>
Date: Wednesday, March 5, 2025 at 10:02 AM
To: lamps-wg/dilithium-certificates ***@***.***>
Cc: "Massimo, Jake" ***@***.***>, Author ***@***.***>
Subject: Re: [lamps-wg/dilithium-certificates] ML-DSA private key details (PR #87)


This one seems to loose the pretty print. Should we add those back in?

—
Reply to this email directly, view it on GitHub<https://github.com/lamps-wg/dilithium-certificates/pull/87#issuecomment-2701687966>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQHHONA4P2BF2GAAHK46YML2S433JAVCNFSM6AAAAABX6N3ZN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRGY4DOOJWGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
[Image removed by sender. seanturner]seanturner left a comment (lamps-wg/dilithium-certificates#87)<https://github.com/lamps-wg/dilithium-certificates/pull/87#issuecomment-2701687966>

This one seems to loose the pretty print. Should we add those back in?

—
Reply to this email directly, view it on GitHub<https://github.com/lamps-wg/dilithium-certificates/pull/87#issuecomment-2701687966>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AQHHONA4P2BF2GAAHK46YML2S433JAVCNFSM6AAAAABX6N3ZN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBRGY4DOOJWGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

seanturner commented:
@jakemas Are you going to add the section that describes how to generate pk from sk (mentioned in response to Viktor) into this PR or another?

BTW - I checked all 9 PEM examples and they look right to me.

jakemas commented:
Thanks for the review! As the `pk` from `sk` section will be independent of this, and this is already a sizable PR, I will split into a new PR.
